RealmSwift/Map.swift

```diff
@@ -0,0 +1,600 @@
+//
+// Copyright 2021 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+import Foundation
+import Realm
+import Realm.Private
+
+/// :nodoc:
+public protocol _MapKey: Hashable {
+    static var _rlmType: RLMPropertyType { get }
+}
+extension String: _MapKey { }
+
+/**
+ Map is a key-value storage container used to store supported Realm types.
+ 
+ Map is a generic type that is parameterized on the type it stores. This can be either an Object
+ subclass or one of the following types: Bool, Int, Int8, Int16, Int32, Int64, Float, Double,
+ String, Data, Date, Decimal128, and ObjectId (and their optional versions)
+ 
+ Map only supports String as a key.
+ 
+ Unlike Swift's native collections, `Map`s is a reference types, and are only immutable if the Realm that manages them
+ is opened as read-only.
+ 
+ A Map can be filtered and sorted with the same predicates as `Results<Value>`.
+ 
+ Properties of `Map` type defined on `Object` subclasses must be declared as `let` and cannot be `dynamic`.
+*/
+public final class Map<Key, Value>: RLMSwiftCollectionBase where Key: _MapKey, Value: RealmCollectionValue {
+
+    // MARK: Properties
+
+    /// The Realm which manages the map, or `nil` if the map is unmanaged.
+    public var realm: Realm? {
+        return _rlmCollection.realm.map { Realm($0) }
+    }
+
+    /// Indicates if the map can no longer be accessed.
+    public var isInvalidated: Bool { return _rlmCollection.isInvalidated }
+
+    /// Returns all of the keys in this map.
+    public var keys: [Key] {
+        return rlmDictionary.allKeys.map(dynamicBridgeCast)
+    }
+
+    /// Returns all of the values in this map.
+    public var values: [Value] {
+        return rlmDictionary.allValues.map(dynamicBridgeCast)
+    }
+
+    // MARK: Initializers
+
+    /// Creates a `Map` that holds Realm model objects of type `Value`.
+    public override init() {
+        super.init()
+    }
+
+    internal init(objc rlmDictionary: RLMDictionary<AnyObject, AnyObject>) {
+        super.init(collection: rlmDictionary)
+    }
+
+    // MARK: Count
+
+    /// Returns the number of key-value pairs in this map.
+    @objc public var count: Int { return Int(_rlmCollection.count) }
+
+    // MARK: Mutation
+
+    /**
+     Updates the value stored in the map for the given key, or adds a new key-value pair if the key does not exist.
```

Comment:
I noted some things in the docs for RealmKeyedCollection that apply to this file as well.